### ISSUE TITLE
The implementation of gradient accumulation.

### ISSUE THREAD

**Issue by hlnull** [swookey-thinky/xdiffusion#8]
Hi, I found the implementation of gradient accumulation might be a little confused.

The accelerator has been initialized with GradientAccumulationPlugin(num_steps=gradient_accumulation_steps). 
```
    accelerator = Accelerator(
        dataloader_config=DataLoaderConfiguration(split_batches=False),
        mixed_precision=mixed_precision,
        gradient_accumulation_plugin=(
            GradientAccumulationPlugin(
                num_steps=gradient_accumulation_steps,
                adjust_scheduler=True,
                sync_with_dataloader=False,
            )
            if gradient_accumulation_steps > 1
            else None
        ),
        kwargs_handlers=[ddp_kwargs],
        step_scheduler_with_optimizer=False,
        cpu=accelerator_force_cpu,
    )
```

Given this setup, is the line for _ in range(gradient_accumulation_steps) still necessary? Since accelerator.accumulate(vae) is designed to handle gradient accumulation, it seems the explicit loop might be redundant. 
Am I correct in thinking this way?

```
# All of the gradient accumulation steps count as one training step.
for _ in range(gradient_accumulation_steps):
  with accelerator.accumulate(vae):
    for optimizer_idx, optimizer in enumerate(optimizers):
      with accelerator.autocast():
        loss, .. = vae()

    # calculate the gradients at each step in the network.
    accelerator.backward(losses)

    for optimizer in enumerate(optimizers):
      optimizer.step()
      optimizer.zero_grad()
```

**Comment by swookey-thinky** [comment]
The loop over the gradient accumulation steps is only there to keep the meaning of `num_training_steps` consistent. In the code, `num_training_steps` is equivalent to "the number of backpropagation steps", e.g. the number of times the optimizer updates the weights of the model. Suppose you have `num_training_steps=100` and `gradient_accumulation_steps=1` (e.g. simple training without gradient accumulation). In this scenario, the optimizer will apply 100 updates to the model, according to `num_training_steps`. Now suppose you have `num_training_steps=100` but `gradient_accumulation_steps=4`. In this case, without that extra loop above, the optimizer will only update the model 25 times (e.g. `num_training_steps / gradient_accumulation_steps`). So to get the same number of model updates, you would need to multiply `num_training_steps` by 4. I decided to keep `num_training_steps` consistent across `gradient_accumulation_steps` settings, but honestly I can see the argument both ways. It made sense when I was the only one using it, but open to other ideas :)

**Comment by hlnull** [comment]
Apologize for the late response. Mondays are always just too tough :(.

> Now suppose you have num_training_steps=100 but gradient_accumulation_steps=4. In this case, without that extra loop above, the optimizer will only update the model 25 times (e.g. num_training_steps / gradient_accumulation_steps). So to get the same number of model updates, you would need to multiply num_training_steps by 4.

You're absolutely right. If the user sets gradient_accumulation_steps to 4, they should multiply num_training_steps by 4 accordingly.

I now see the differences in the implementation. Thanks for clarifying that!
